kubernetes/resource_kubernetes_replication_controller.go

```diff
@@ -0,0 +1,258 @@
+	spec, err := expandReplicationControllerSpec(d.Get("spec").([]interface{}))
+	if err != nil {
+		return err
+	}
+
+	spec.Template.Spec.AutomountServiceAccountToken = ptrToBool(false)
+
+	rc := api.ReplicationController{
+		ObjectMeta: metadata,
+		Spec:       spec,
+	}
+
+	log.Printf("[INFO] Creating new replication controller: %#v", rc)
+	out, err := conn.CoreV1().ReplicationControllers(metadata.Namespace).Create(&rc)
+	if err != nil {
+		return fmt.Errorf("Failed to create replication controller: %s", err)
+	}
+
+	d.SetId(buildId(out.ObjectMeta))
+
+	log.Printf("[DEBUG] Waiting for replication controller %s to schedule %d replicas",
+		d.Id(), *out.Spec.Replicas)
+	// 10 mins should be sufficient for scheduling ~10k replicas
+	err = resource.Retry(d.Timeout(schema.TimeoutCreate),
+		waitForDesiredReplicasFunc(conn, out.GetNamespace(), out.GetName()))
+	if err != nil {
+		return err
+	}
+	// We could wait for all pods to actually reach Ready state
+	// but that means checking each pod status separately (which can be expensive at scale)
+	// as there's no aggregate data available from the API
+
+	log.Printf("[INFO] Submitted new replication controller: %#v", out)
+
+	return resourceKubernetesReplicationControllerRead(d, meta)
+}
+
+func resourceKubernetesReplicationControllerRead(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*kubernetes.Clientset)
+
+	namespace, name := idParts(d.Id())
+	log.Printf("[INFO] Reading replication controller %s", name)
+	rc, err := conn.CoreV1().ReplicationControllers(namespace).Get(name, metav1.GetOptions{})
+	if err != nil {
+		log.Printf("[DEBUG] Received error: %#v", err)
+		return err
+	}
+	log.Printf("[INFO] Received replication controller: %#v", rc)
+
+	err = d.Set("metadata", flattenMetadata(rc.ObjectMeta))
+	if err != nil {
+		return err
+	}
+
+	spec, err := flattenReplicationControllerSpec(rc.Spec)
+	if err != nil {
+		return err
+	}
+
+	err = d.Set("spec", spec)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func resourceKubernetesReplicationControllerUpdate(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*kubernetes.Clientset)
+
+	namespace, name := idParts(d.Id())
+
+	ops := patchMetadata("metadata.0.", "/metadata/", d)
+
+	if d.HasChange("spec") {
+		spec, err := expandReplicationControllerSpec(d.Get("spec").([]interface{}))
+		if err != nil {
+			return err
+		}
+
+		ops = append(ops, &ReplaceOperation{
+			Path:  "/spec",
+			Value: spec,
+		})
+	}
+	data, err := ops.MarshalJSON()
+	if err != nil {
+		return fmt.Errorf("Failed to marshal update operations: %s", err)
+	}
+	log.Printf("[INFO] Updating replication controller %q: %v", name, string(data))
+	out, err := conn.CoreV1().ReplicationControllers(namespace).Patch(name, pkgApi.JSONPatchType, data)
+	if err != nil {
+		return fmt.Errorf("Failed to update replication controller: %s", err)
+	}
+	log.Printf("[INFO] Submitted updated replication controller: %#v", out)
+	d.SetId(buildId(out.ObjectMeta))
```

Comment:
Is it normal for the ID of the resource to change here? Does the old ID still work? Is it possible for the `Patch` operation above (~line 170) to encounter an error but still actually change the id, in which case something is lost?

If this kind of id updating is normal then ignore me. I've seen other resources re-set the ID in UPDATE methods when it wasn't necessary (or recommended)